decisions/store-preferences.md

```diff
@@ -0,0 +1,171 @@
+# Store Preferences
+
+- *Date*: 
+- *Deciders*:
+- *Status*: 
+- *Outcome*:
+- *Related Material*: issue comment [Option 1](https://github.com/openmsupply/open-msupply/issues/973#issuecomment-1401165427), [Option 2](https://github.com/openmsupply/open-msupply/issues/973#issuecomment-1397753763) 
+
+## Background
+
+Preferences are used to configure behaviour of application, there are multiple preferences in mSupply. 
+Preferences apply to different areas (i.e. by store, globally, etc..). 
+Preferences in mSupply are stored in [pref] table with preference values in [pref]data and 'area' where they apply in other columns i.e. [pref]store_id. 
+This KDD is focused on Store Preferences but a wider preferences context should be considered, see Appending at the bottom of KDD for an example of store preference record in mSupply.
+
+## Preferences and sync
+
+[pref] table does sync, it's not straightforward though, as [pref] table can be central, remote and local data. 
+Store Preferences are centrally controlled and will sync to corresponding stores
+
+## Requirements
+
+1. Sync, store and use preferences for a store
+2. Store will have multiple preferences, each of different type
+3. Preferences are known at compile time
+4. Preference for store may not exist (should have default value)
+
+## Related or Future requirements
+
+Global preferences -> Wouldn't need store_id
+Local preferences -> Similar to key value store, but for local store preferences would need to add store_id to key value store
+Dynamic preferences -> plugins ?
+
+## Options
+
+### Option 1 - [preference] -> [preference_store_join]
+
+Use [preference] and [preference_store_join] table
+
+The idea is taken from [this comment issue](https://github.com/openmsupply/open-msupply/issues/973#issuecomment-1401165427).
+
+[preference] table is populated with preference types and values are store on [preference_store_join], [preference_store_join] would need to have optional value columns of all of the different types (similar to key value store)
+
+*Pros:*
+
+- Faster to add new preference
+
+*Cons:*
+
+- Schema does not describe what preferences are available
+- Harder for new devs to understand
+- Value typing is not enforced by schema
+- More tables to sync
+- Schema does not restrict one preference value for store (needs to be guaranteed in code)
+
+### Option 2 - Table
+
+[store_preference] table, [store_preference]store_id as primary key and strongly typed columns for preferences for that store. 
+StorePreferenceRow data type specifies default values with default trait implementation unwrap_or_default can be used if store_preference does not exist for a store.
```

Comment:
Is there a need allow nullable preferences, e.g. if a store hasn't had a specific preference set do we need to allow it to stay as unset, or are we ok to always inherit the default preference (as if it was persisted to the database)

I can't think if a great example right now, but I'm thinking there might be situations where `unwrap_or_default` might still result in some preferences needing to be checked for `None`.

Not really a problem but thought it might be worth considering.